mscore/scoreview.cpp

```diff
@@ -5494,6 +5495,7 @@ void ScoreView::cmdAddAnnotation2()
       annotation->setTrack(cr->track());
       annotation->textAnnotation()->setTextStyleType(TextStyleType::ANNOTATION);
       annotation->setParent(cr->segment());
+      cr->segment()->add(annotation);
```

Comment:
I'd commented undoAddElement earlier when it was causing a crash (fixed it
later). You were right, on adding it back the layout is being called for
annotation element (due to the Segment::add call) after also being added to
the segment annotation list. It still isn't being drawn yet though. Maybe
something similar needs to be done for draw as well (explicit call).

On Mon, Jun 20, 2016 at 7:49 PM, Marc Sabatella <notifications@github.com>
wrote:

> In mscore/scoreview.cpp
> <https://github.com/musescore/MuseScore/pull/2666#discussion_r67697242>:
>
> > @@ -5494,6 +5495,7 @@ void ScoreView::cmdAddAnnotation2()
> >        annotation->setTrack(cr->track());
> >        annotation->textAnnotation()->setTextStyleType(TextStyleType::ANNOTATION);
> >        annotation->setParent(cr->segment());
> > +      cr->segment()->add(annotation);
>
> Look to see how other elements are added. We seldom if ever call
> Segment::add() directly; it is usually Score::undoAddElement(). Look at
> ScoreView::cmdAddText() in particular for a model of how this is done.
> Doing it this way is necessary to get things to work with linked parts and
> also with undo. You will probably also need to add handling for your new
> element type within Score::undoAddElement(). Again, look to see how staff
> text is handled, and start by assuming a segment-attached annotation will
> need to do the same things. The range-attached annotations will be
> considerably trickier, but we'll cross that bridge when we get there.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/musescore/MuseScore/pull/2666/files/961c945ad275e097849fe3808bcd3f32e5cc14d3..3e62a0086919fe87f5c39c182d95639b70694c5a#r67697242>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGf7yt2ZMnxyJHU23L7d_gTwKdZLuNyJks5qNqFzgaJpZM4I1g-h>
> .
>



-- 
Ruchit Agrawal.
IIIT Hyderabad.